include/sys/arch_interface.h

```diff
@@ -0,0 +1,395 @@
+ * @param prio Thread priority.
+ * @param options Thread options.
+ */
+void z_arch_new_thread(struct k_thread *thread, k_thread_stack_t *pStack,
+		       size_t stackSize, k_thread_entry_t entry,
+		       void *p1, void *p2, void *p3,
+		       int prio, unsigned int options);
+
+#ifdef CONFIG_USE_SWITCH
+/**
+ * Cooperatively context switch between two threads.
+ *
+ * The semantics of the two switch handles are entirely up to the architecture
+ * implementation, and are stored per thread in k_thread->switch_handle.
+ * The outgoing thread's switch handle pointer may be updated during this
+ * process.
+ *
+ * @param switch_to Incoming thread's switch handle
+ * @param switched_from Pointer to outgoing thread's switch handle, which
+ *		may be updated.
+ */
+void z_arch_switch(void *switch_to, void **switched_from);
+#else
+int z_arch_swap(unsigned int key);
+#endif
+
+#ifndef CONFIG_USE_SWITCH
+/** Set the return value for when the thread is swapped in again */
+static ALWAYS_INLINE void
+z_arch_thread_return_value_set(struct k_thread *thread, unsigned int value);
+#endif
+
+#ifdef CONFIG_ARCH_HAS_CUSTOM_SWAP_TO_MAIN
+/**
+ * Custom logic for entering main thread context at early boot
+ *
+ * Used by architectures for whom the typical trick of setting up
+ * a dummy thread in early boot context to "switch out" of doesn't
+ * isn't workable.
+ *
+ * @param main_thread main thread object
+ * @param main_stack main thread's stack object
+ * @param main_stack_size Size of the stack object's buffer
+ * @param _main Entry point for application main function.
+ */
+void z_arch_switch_to_main_thread(struct k_thread *main_thread,
+				  k_thread_stack_t *main_stack,
+				  size_t main_stack_size,
+				  k_thread_entry_t _main);
+#endif
+
+#if defined(CONFIG_FLOAT) && defined(CONFIG_FP_SHARING)
+/**
+ * @brief Disable floating point context preservation
+ *
+ * The function is used to disable the preservation of floating
+ * point context information for a particular thread.
+ *
+ * @note
+ * For ARM architecture, disabling floating point preservation
+ * - may only be requested for the current thread
+ * - cannot be requested in ISRs.
+ *
+ * @retval 0       On success.
+ * @retval -EINVAL If the floating point disabling could not be performed.
+ */
+int z_arch_float_disable(struct k_thread *thread);
+#endif /* CONFIG_FLOAT && CONFIG_FP_SHARING */
+
+/** @} */
+
+/**
+ * @defgroup arch-pm Architecture-specific power management APIs
+ * @{
+ */
+/** Halt the system, optionally propagating a reason code */
+FUNC_NORETURN void z_arch_system_halt(unsigned int reason);
+
+/** @} */
+
+
+/**
+ * @defgroup arch-smp Architecture-specific SMP APIs
+ * @{
+ */
+#ifdef CONFIG_SMP
+/** Return the CPU struct for the currently executing CPU */
+static inline struct _cpu *z_arch_curr_cpu(void);
+#endif
+
+/**
+ * Broadcast an interrupt to all CPUs
+ *
+ * This will invoke z_sched_ipi() on other CPUs in the system.
+ */
+void z_arch_sched_ipi(void);
```

Comment:
should this one not be under the CONFIG_SMP ifdef?

Reply:
fixed